libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,363 @@
+				return true;
+			}
+
+		// At this point we want to keep all slots.
+		for (size_t i = 0; i < ops.sourceSize(); ++i)
+			yulAssert(ops.sourceMultiplicity(i) >= 0, "");
+		yulAssert(ops.sourceSize() <= ops.targetSize(), "");
+
+		// If the top is not in position, try to find a slot that wants to be at the top and swap it up.
+		if (!ops.isCompatible(sourceTop, sourceTop))
+			for (size_t sourceOffset: ranges::views::iota(0u, ops.sourceSize()))
+				if (
+					!ops.isCompatible(sourceOffset, sourceOffset) &&
+					ops.isCompatible(sourceOffset, sourceTop)
+				)
+				{
+					ops.swap(ops.sourceSize() - sourceOffset - 1);
+					return true;
+				}
+
+		// If we still need more slots, produce a suitable one.
+		if (ops.sourceSize() < ops.targetSize())
+		{
+			bringUpTargetSlot(ops, ops.sourceSize());
+			return true;
+		}
+
+		// The stack has the correct size, each slot has the correct number of copies and the top is in position.
+		yulAssert(ops.sourceSize() == ops.targetSize(), "");
+		size_t size = ops.sourceSize();
+		for (size_t i = 0; i < ops.sourceSize(); ++i)
+			yulAssert(ops.sourceMultiplicity(i) == 0 && (ops.targetIsArbitrary(i) || ops.targetMultiplicity(i) == 0), "");
+		yulAssert(ops.isCompatible(sourceTop, sourceTop), "");
+
+		// If we find a lower slot that is out of position, but also compatible with the top, swap that up.
+		for (size_t offset: ranges::views::iota(0u, size))
+			if (!ops.isCompatible(offset, offset) && ops.isCompatible(sourceTop, offset))
+			{
+				ops.swap(size - offset - 1);
+				return true;
+			}
+		// Swap up any slot that is still out of position.
+		for (size_t offset: ranges::views::iota(0u, size))
+			if (!ops.isCompatible(offset, offset) && !ops.sourceIsSame(offset, sourceTop))
+			{
+				ops.swap(size - offset - 1);
+				return true;
+			}
+		yulAssert(false, "");
+	}
+};
+
+
+/// Transforms @a _currentStack to @a _targetStack, invoking the provided shuffling operations.
+/// Modifies @a _currentStack itself after each invocation of the shuffling operations.
+template<typename Swap, typename PushOrDup, typename Pop>
+void createStackLayout(Stack& _currentStack, Stack const& _targetStack, Swap _swap, PushOrDup _pushOrDup, Pop _pop)
+{
+	struct ShuffleOperations
+	{
+		Stack& currentStack;
+		Stack const& targetStack;
+		Swap swapCallback;
+		PushOrDup pushOrDupCallback;
+		Pop popCallback;
+		std::map<StackSlot, int> multiplicity;
+		ShuffleOperations(
+			Stack& _currentStack,
+			Stack const& _targetStack,
+			Swap _swap,
+			PushOrDup _pushOrDup,
+			Pop _pop
+		):
+			currentStack(_currentStack),
+			targetStack(_targetStack),
+			swapCallback(_swap),
+			pushOrDupCallback(_pushOrDup),
+			popCallback(_pop)
+		{
+			for (auto const& slot: currentStack)
+				--multiplicity[slot];
+			for (auto&& [offset, slot]: targetStack | ranges::views::enumerate)
+				if (std::holds_alternative<JunkSlot>(slot) && offset < currentStack.size())
+					++multiplicity[currentStack.at(offset)];
```

Comment:
Is this just incrementing the multiplicity of a more or less arbitrary source slot?

Reply:
No, the multiplicity at the slot that currently is at that target position. Whatever we have in the source at that offset can just stay there, since we can have whatever we want in the target at that position - so there is no reason to change the slot, i.e. we try to just keep whatever happens to be at that position in the source. But that means that all in all I need one more copy of whatever happens to be at that position than I would otherwise expect.

Reply:
Or well, I guess in that sense one could say it is a "more or less arbitrary" slot, but not without reason :-).

Reply:
But is that really optimal? If you have `1 2 3` -> `3 2 ?`, then requiring a multiplicity of `1` for `3` does not lead to the optimal sequence, does it?

Reply:
It will use the optimal sequence for that, but yeah, the multiplicity for ``3`` will be "wrong"/misleading in that case... this is only an issue, if the JunkSlot is at the stack top - that's the reason for ``|| ops.targetIsArbitrary(sourceTop)`` in L225, which is among the very first things that happen. In the example that rule will immediately move the 3 in place and not care about any multiplicities or the fact that the target slot is arbitrary (the idea being: it's always optimal to move the stack top, whenever there is anything to gain by it, in this case finalizing the stack bottom).

Whenever the junk slot is *not* at the top, I don't ever want to touch or move it, though, since there's nothing to gain by changing anything about it, so it'd just be a wasted operation. And only in those cases the multiplicities start to be considered at all.

Reply:
Or well, it's complicated :-)... e.g. in ``1 2 3`` -> ``1 1 ?`` that rule *won't* kick in... I actually have to think through what happens there... An ideal way to do this would be ``POP DUP2 SWAP1``, I guess, if we wouldn't force the multiplicity of 3 to be 1.

What ``shuffleStep`` will currently do is... the first rule to kick in is L240+... which will dup the 1, i.e. DUP3. Then we have ``1 2 3 1``, so it will swap the 1 into place, i.e. SWAP2  to ``1 1 3 2`` and the top is no longer needed, i.e. POP, so ``DUP3 SWAP2 POP``... so works out in this case :-). No idea if it always will, though :-).

But yeah, I don't know, it seems to work in any case and I'd say it's good enough for now - if it's correct and safe, then we can still try to tweak it later.